query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+                // If the value is known to be non-textual, we can simply use the regular datoms
+                // table (TODO: and exclude on `index_fulltext`!).
+                //
+                // If the value is a placeholder too, then we can walk the non-value-joined view,
+                // because we don't care about retrieving the fulltext value.
+                //
+                // If the value is a variable or string, we must use `all_datoms`, or do the join
+                // ourselves, because we'll need to either extract or compare on the string.
+                Some(
+                    match value {
+                        // TODO: see if the variable is projected, aggregated, or compared elsewhere in
+                        // the query. If it's not, we don't need to use all_datoms here.
+                        &PatternValuePlace::Variable(_) =>
+                            DatomsTable::AllDatoms,       // TODO: check types.
+                        &PatternValuePlace::Constant(NonIntegerConstant::Text(_)) =>
+                            DatomsTable::AllDatoms,
+                        _ =>
+                            DatomsTable::Datoms,
+                    }),
+            &PatternNonValuePlace::Entid(id) =>
+                schema.attribute_for_entid(id)
+                      .and_then(|attribute| self.table_for_attribute_and_value(attribute, value)),
+            &PatternNonValuePlace::Ident(ref kw) =>
+                self.attribute_for_ident(schema, kw)
+                    .and_then(|attribute| self.table_for_attribute_and_value(attribute, value)),
+        }
+    }
+
+    /// Produce a (table, alias) pair to handle the provided pattern.
+    /// This is a mutating method because it mutates the aliaser function!
+    /// Note that if this function decides that a pattern cannot match, it will flip
+    /// `is_known_empty`.
+    fn alias_table<'s, 'a>(&mut self, schema: &'s Schema, pattern: &'a Pattern) -> Option<SourceAlias> {
+        self.table_for_places(schema, &pattern.attribute, &pattern.value)
+            .map(|table| SourceAlias(table, (self.aliaser)(table)))
+    }
+
+    fn get_attribute<'s, 'a>(&self, schema: &'s Schema, pattern: &'a Pattern) -> Option<&'s Attribute> {
+        match pattern.attribute {
+            PatternNonValuePlace::Entid(id) =>
+                schema.attribute_for_entid(id),
+            PatternNonValuePlace::Ident(ref kw) =>
+                self.attribute_for_ident(schema, kw),
+            _ =>
+                None,
+        }
+    }
+
+    fn get_value_type<'s, 'a>(&self, schema: &'s Schema, pattern: &'a Pattern) -> Option<ValueType> {
+        self.get_attribute(schema, pattern).map(|x| x.value_type)
+    }
+}
+
+/// Expansions.
+impl ConjoiningClauses {
+
+    /// Take the contents of `bindings` and generate inter-constraints for the appropriate
+    /// columns into `wheres`.
+    /// For example, a bindings map like
```

Comment:
Heh.  Best to bump comments; they're good for me (and easy for you) but not valuable for others.